atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresGetCandidateCellsForSweeping.java

```diff
@@ -263,59 +190,42 @@ private FullQuery getQuery() {
-                    + "      WHERE ts < ? " + bounds.predicates + getIgnoredTimestampPredicate()
-                    + "      ORDER BY row_name, col_name, ts"
-                    + "      LIMIT " + sqlRowLimit
-                    + "    ) sub"
-                    + "    GROUP BY row_name, col_name"
-                    + "    ORDER BY row_name, col_name"
-                    + "  ) cells"
-                    +    (request.shouldCheckIfLatestValueIsEmpty()
-                            ? "  JOIN " + prefixedTableName + " v"
-                              + "  ON cells.row_name = v.row_name"
-                              + "  AND cells.col_name = v.col_name"
-                              + "  AND cells.max_ts = v.ts"
-                            : "")
-                    + "  WHERE"
-                      // See KVS.getCandidateCellsForSweeping() docs for the definition of a candidate cell:
-                      // (1) The set T has more than one element
-                    + "    min_ts <> max_ts"
-                      // (2) The set T contains an element that is greater than or equal to Tu
-                    + "    OR max_ts >= ?"
-                      // (3) The set T contains Value.INVALID_VALUE_TIMESTAMP
-                    +      (ignoreSentinels ? "" : " OR have_sentinel = 1")
-                      // (4) V is true and the cell value corresponding to the maximum element of T is empty
-                    +      (request.shouldCheckIfLatestValueIsEmpty() ? " OR length(v.val) = 0" : "")
-                      // Also, always get the first cell, as well as the last one if the limit was reached
-                    + "    OR min_rn = 1 OR max_rn = " + sqlRowLimit
-                    + "  ORDER BY cells.row_name, cells.col_name";
-            return new FullQuery(query)
-                    .withArg(request.sweepTimestamp()) // "WHERE ts < ?"
-                    .withArgs(bounds.args)
-                    .withArg(request.minUncommittedStartTimestamp()); // "OR max_ts >= ?"
-        }
-
-        private boolean areSentinelsIgnored() {
-            return Longs.contains(request.timestampsToIgnore(), Value.INVALID_VALUE_TIMESTAMP);
+            if (request.shouldCheckIfLatestValueIsEmpty()) {
+                String query = "/* GET_CANDIDATE_CELLS_FOR_SWEEPING_THOROUGH(" + tableName + ") */"
+                        + "  SELECT cells.row_name, cells.col_name, cells.timestamps, "
+                        + "         length(v.val) = 0 AS latest_val_empty"
+                        + "  FROM ("
+                        + "    SELECT"
+                        + "      row_name, col_name, MAX(ts) AS max_ts, ARRAY_AGG(ts) AS timestamps"
+                        + "    FROM ("
+                        + "      SELECT row_name, col_name, ts"
+                        + "      FROM " + prefixedTableName
+                        + "      WHERE ts < ? " + bounds.predicates + getIgnoredTimestampPredicate()
```

Comment:
should thorough have the getIgnoredTimestampPredicate? This brings us to the question if `shouldCheckIfLatestValueIsEmpty`  and `timestampsToIgnore` should be two separate values or grouped as SweepStrategyParams (maybe an enum with types CONSERVATIVE/ThOROUGH).

timestampsToIgnore will be -1 for Conservative and empty for THOROUGH and we can add binds in the conservative query only..

Reply:
We can discuss what the API should be (and in fact we could have discussed this a few months back when I introduced it), but in this PR I'm trying to keep changes minimal.

This is a question of the level of abstraction. In the current design, KVS doesn't really know anything about sweep strategies; we just tell it whether to ignore timestamps and/or load empty values. In this design, the two concepts are orthogonal, and that's why you see getIgnoredTimestampPredicate() in both "if" branches. We could change the design and make KVS aware of sweep strategies, but that's beyond the scope of this pull request.

Reply:
I am just pointing out that `shouldCheckIfLatestValueIsEmpty` and `timestampsToIgnore` are not independent concepts but are correlated based on the sweep strategy, seems like a bad design to allow many more cases here when there are exactly two possibilities.

Reply:
In general, I think that the KVS should not know about sweep - its one job is to make the appropriate calls to the DB (via the clientPool, in the C* case) in a correct and efficient way. It shouldn't have to know why the user wants to get these timestamps - +1 to keeping as-is.

However, the boundaries are blurred somewhat as we're specifically satisfying a `getCandidateCellsForSweeping` request, implying that it's reasonable to know something about what being a CandidateCellForSweeping might entail.

But also, the branching here is no more than it was before this PR, and I don't think it's an egregious combinatorial explosion.

On balance, I'd like this to merge as is - any refactor to try and cut down on the branching should be in a separate PR, and should also be a low priority, given the limited potential for simplification and the urgency of getting this merged in.

Reply:
Regarding removing branching - I want to benchmark a different query later that would eliminate this branching (for thorough, check values for all rows rather than only load the latest values). But of course I need to get the Oracle part done first - we can get back to this later.